cloudpathlib/cloudpath.py

```diff
@@ -609,6 +611,90 @@ def rmtree(self):
             )
         self.client._remove(self)
 
+    def upload_from(
+        self, source: Union[str, os.PathLike], force_overwrite_to_cloud: bool = False
+    ) -> "CloudPath":
+        """Upload a file or directory to the cloud path."""
+        source = Path(source)
+
+        if source.is_dir():
+            for p in source.iterdir():
+                (self / p.name).upload_from(p, force_overwrite_to_cloud=force_overwrite_to_cloud)
+
+        else:
+            if self.exists() and self.is_dir():
+                dst = self / source.name
+            else:
+                dst = self
+
+            dst._upload_file_to_cloud(source, force_overwrite_to_cloud=force_overwrite_to_cloud)
+
+        return self
+
+    def copy(
+        self,
+        destination: Union[str, os.PathLike, "CloudPath"],
+        force_overwrite_to_cloud: bool = False,
+    ) -> Union[str, os.PathLike, "CloudPath"]:
+        """Copy self to destination folder of file, if self is a file."""
+        if not self.exists() or not self.is_file():
+            raise ValueError(
+                f"Path {self} should be a file. To copy a directory tree use the method copytree."
+            )
+
+        if not isinstance(destination, CloudPath):
+            return self.download_to(destination)
+
+        # if same client, use cloud-native _move_file on client to avoid downloading
+        elif self.client is destination.client:
+            if (
+                not force_overwrite_to_cloud
+                and destination.exists()
+                and destination.stat().st_mtime >= self.stat().st_mtime
+            ):
+                raise OverwriteNewerCloudError(
+                    f"File ({destination}) is newer than ({self}). "
+                    f"To overwrite "
+                    f"pass `force_overwrite_to_cloud=True`."
+                )
+
+            return self.client._move_file(self, destination, remove_src=False)
+
+        else:
+            if not destination.exists() or destination.is_file():
+                return destination.upload_from(
+                    self.fspath, force_overwrite_to_cloud=force_overwrite_to_cloud
+                )
+            else:
+                return (destination / self.name).upload_from(
+                    self.fspath, force_overwrite_to_cloud=force_overwrite_to_cloud
+                )
+
+    def copytree(
+        self, destination: "CloudPath", force_overwrite_to_cloud: bool = False
+    ) -> Union[os.PathLike, "CloudPath"]:
+        """Copy self to a directory, if self is a directory."""
+        if not self.is_dir():
+            raise ValueError(
+                f"Origin path {self} must be a directory. To copy a single file use the method copy."
+            )
+        if destination.exists() and destination.is_file():
+            raise ValueError("Destination path {destination} of copytree must be a directory.")
```

Comment:
I think we want to use [`CloudPathFileExistsError`](https://cloudpathlib.drivendata.org/api-reference/exceptions/#cloudpathlib.exceptions.CloudPathFileExistsError).